File: sparkcql-core/src/main/scala/gov/va/sparkcql/adapter/library/LibraryAdapterAggregator.scala
package gov.va.sparkcql.adapter.library

import org.hl7.elm.r1.VersionedIdentifier

sealed class LibraryAdapterAggregator(libraryAdapters: List[LibraryAdapter]) extends LibraryAdapter {
  
  def getLibraryContent(identifier: VersionedIdentifier): Option[String] = {
    val found = libraryAdapters.flatMap(_.getLibraryContent(identifier))
    found.headOption
  }
}

File: sparkcql-core/src/main/java/gov/va/sparkcql/adapter/library/LibraryAdapterAggregator.java
package gov.va.sparkcql.adapter.library;

import java.util.List;

import org.hl7.elm.r1.VersionedIdentifier;

public class LibraryAdapterAggregator implements LibraryAdapter {
    
    protected List<LibraryAdapter> libraryAdapters;

    public LibraryAdapterAggregator(List<LibraryAdapter> libraryAdapters) {
        this.libraryAdapters = libraryAdapters;
    }

    @Override
    public String getLibrarySource(VersionedIdentifier identifier) {
        var resolvedAdapter = libraryAdapters.stream().filter(a -> a.containsLibrary(identifier));
        if (resolvedAdapter.count() > 0) {
            return resolvedAdapter.findFirst().get().getLibrarySource(identifier);
        } else {
            throw new RuntimeException("Library " + identifier.toString() + " not found.");
        }
    }

    @Override
    public Boolean containsLibrary(VersionedIdentifier identifier) {
        return libraryAdapters.stream().filter(a -> a.containsLibrary(identifier)).count() > 0;
    }
}